File: common/src/client/java/juuxel/adorn/client/gui/widget/PageContainer.java
package juuxel.adorn.client.gui.widget;

public interface PageContainer<T> {
    // 0-indexed
    int getCurrentPage();
    void setCurrentPage(int currentPage);

    int getPageCount();
    T getCurrentPageValue();

    default boolean hasPreviousPage() {
        return getCurrentPage() > 0;
    }

    default boolean hasNextPage() {
        return getCurrentPage() < getPageCount() - 1;
    }

    void showPreviousPage();
    void showNextPage();
}

File: common/src/main/kotlin/juuxel/adorn/client/gui/widget/PageContainer.kt
package juuxel.adorn.client.gui.widget

interface PageContainer<T> {
    var currentPage: Int // 0-indexed
    val pageCount: Int
    val currentPageValue: T

    fun hasPreviousPage(): Boolean = currentPage > 0
    fun hasNextPage(): Boolean = currentPage < pageCount - 1

    fun showPreviousPage()
    fun showNextPage()

    fun showPage(value: T)
}
